homeassistant/components/roomba/config_flow.py

```diff
@@ -0,0 +1,127 @@
+"""Config flow to configure roomba component."""
+import logging
+
+from roomba import Roomba
+import voluptuous as vol
+
+from homeassistant import config_entries, core
+from homeassistant.const import CONF_HOST, CONF_PASSWORD, CONF_USERNAME
+from homeassistant.core import callback
+
+from . import (
+    CannotConnect,
+    async_connect_or_timeout as act,
+    async_disconnect_or_timeout as adt,
+)
+from .const import (
+    CONF_CERT,
+    CONF_CONTINUOUS,
+    CONF_DELAY,
+    DEFAULT_CERT,
+    DEFAULT_CONTINUOUS,
+    DEFAULT_DELAY,
+)
+from .const import DOMAIN  # pylint:disable=unused-import
+
+DATA_SCHEMA = vol.Schema(
+    {
+        vol.Required(CONF_HOST): str,
+        vol.Required(CONF_USERNAME): str,
+        vol.Required(CONF_PASSWORD): str,
+        vol.Optional(CONF_CERT, default=DEFAULT_CERT): str,
+        vol.Optional(CONF_CONTINUOUS, default=DEFAULT_CONTINUOUS): bool,
+        vol.Optional(CONF_DELAY, default=DEFAULT_DELAY): int,
+    }
+)
+
+_LOGGER = logging.getLogger(__name__)
+
+
+async def validate_input(hass: core.HomeAssistant, data):
+    """Validate the user input allows us to connect.
+
+    Data has the keys from DATA_SCHEMA with values provided by the user.
+    """
+    roomba = Roomba(
+        address=data[CONF_HOST],
+        blid=data[CONF_USERNAME],
+        password=data[CONF_PASSWORD],
+        cert_name=data[CONF_CERT],
+        continuous=data[CONF_CONTINUOUS],
+        delay=data[CONF_DELAY],
+    )
+    return await act(hass, roomba)
+
+
+class RoombaConfigFlow(config_entries.ConfigFlow, domain=DOMAIN):
+    """Roomba configuration flow."""
+
+    VERSION = 1
+    CONNECTION_CLASS = config_entries.CONN_CLASS_LOCAL_PUSH
+
+    @staticmethod
+    @callback
+    def async_get_options_flow(config_entry):
+        """Get the options flow for this handler."""
+        return OptionsFlowHandler(config_entry)
+
+    async def async_step_import(self, import_info):
+        """Set the config entry up from yaml."""
+        return await self.async_step_user(import_info)
+
+    async def async_step_user(self, user_input=None):
+        """Handle a flow initialized by the user."""
+        errors = {}
+
+        if user_input is not None:
+            try:
+                info = await validate_input(self.hass, user_input)
+            except CannotConnect:
+                errors = {"base": "cannot_connect"}
+            except Exception:  # pylint: disable=broad-except
+                errors = {"base": "unknown"}
+
+            if "base" not in errors:
+                await adt(self.hass, info["roomba"])
```

Comment:
```suggestion
                await async_disconnect_or_timeout(self.hass, info["roomba"])
```

Ideally we write intention revealing code. It makes it harder for the reviewer if we have to jump all over the file to find what this does.

Reply:
If there is no error, the connection is going to be left open.  This needs to be disconnected in `validate_input`